### DISCUSSION TITLE
Language injection for highlighting punctuation?

### DISCUSSION THREAD

**Discussion by FoamScience** [tree-sitter/tree-sitter#1507] [Q&A] [answered]
Hey guys,

My parser parses relatively complex expressions as a single node, for example:
```
div(1|p,U)  on;
```
will parse as:
```
(identifier) (boolean)
```

Now, for syntax-highlighting the identifier; I'm currently cheating with injections:
```
((identifier) @lua)
```

But that will sometimes generate errors (Because identifiers are not always valid lua statements), or at least do
unnecessary highlighting (eg. all caps variable names will be highlighted by lua parser but I don't want that)

Is there a quick hack to highlight just "punctuation" inside these identifier nodes?

**Comment by alemuller** [comment]
Maybe something like this?

```js
identifier: $ => seq(
  $._id_tok,
  optional(seq(
      token.immediate('('),
      $._ident_expr,
      token.immediate(')'),
  ))
),

// asumming the expr is in the form \w([+,-]\w)*
_ident_expr: $ => choice(
  seq($._id_tok_immd),
  seq($._id_tok_immd, $._op_tok_immd, $._ident_expr),
),

_id_tok:      $ => /[\w]+/,
_id_tok_immd: $ => token.immediate(/[\w]+/),

_op_tok_immd: $ => choice(...['|','+',','].map(token.immediate))
```

Query:
```scm
(identifier
  ["|" "+" ","] @capture)
```

Haven't test, but the concept should work.

**Reply by FoamScience** [reply]
Oh, I'm sorry, I was not informative enough; while your trick works (and I just learned the mapping thing, Thanks),
my identifiers are matched using an external scanner which keeps track of nesting levels of '(' and '<' (It knows
when all parentheses are closed, and marks the end of the identifier on next white space) so the following
```
div( 1|p, U )  on;
```
will still be parsed as:
```
(identifier) (boolean)
```

But, for example:
```
(1 word div( 1|p, U+grad(p)))
```
will parse as:
```
(list (number identifier identifier))
```
The second identifier being: "div( 1|p, U+grad(p))" (Note how it knows that the last ')' belongs to the list)
Hence, this functionality is somewhat important, and I don't want to lose it in favor of highlighting sugar 😄

I thought I would just use another parser to add basic punctuation highlighting but that's too much trouble.
Is there an example of an external parser which can return nested nodes, like this:
```
(identifier (punctuation))
```

**Reply by alemuller** [reply, marked as answer]
>  Is there an example of an external parser which can return nested nodes, like this:

Yes. See #1092.

The example on the linked discussion uses a single external rule. For you use case, more might want to added others rules to `externals`:

```js
externals: $ => [ $._ident_nibble, $.punctuation ],

identifier: $ => repeat(choice(
  $._ident_nibble,
  $.punctuation,
)),
```

Note: Tokens are also allowed as symbols in externals.
```js
externals: $ => [ $._ident_nibble, "|", "+", "grad" ],
```

Since those tokens are hidden from the AST (ie: they are not shown in the output of `tree-sitter parse`). It can be useful to alias them while testing: `alias("|", $.vertical_bar)`.

Going one step further:
```js
externals: $ => [ $.identifier, "+", "|", $._comma_op ],

inline: $ => [ $._operator ],

expression: $ => repeat(choice(
  $.identifier,
  $._operator
)),

_operator: $ => field('operator',choice(
  "+",
  "|",
  // aliasing "," since it is likely already used
  alias($._comma_op, ",")
)),
```

Is there any reason not to implement the expression in the grammar?

**Reply by FoamScience** [reply]
Thank you! this looks a lot better, and actually solves my problem; the externals thing is crazy useful 😄 I just didn't think of using it
